analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,172 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ControllersHaveTooManyResponsibilities : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6960";
+    private const string MessageFormat = "This controller has multiple responsibilities and could be split into {0} smaller units.";
+
+    private static readonly HashSet<string> ExcludedWellKnownServices =
+    [
+        "ILogger",
+        "IMediator",
+        "IMapper",
+        "IConfiguration",
+        "IBus",
+        "IMessageBus"
+    ];
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStartContext =>
+        {
+            if (compilationStartContext.Compilation.ReferencesControllers())
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    var symbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+                    if (symbol.IsControllerType()
+                        && symbol.GetAttributesWithInherited().Any(x => x.AttributeClass.DerivesFrom(KnownType.Microsoft_AspNetCore_Mvc_ApiControllerAttribute)))
+                    {
+                        CheckApiController(symbolStartContext, symbol);
+                    }
+                }, SymbolKind.NamedType);
+            }
+        });
+
+    private static void CheckApiController(SonarSymbolStartAnalysisContext symbolStartContext, INamedTypeSymbol symbol)
+    {
+        var memberNames = RelevantMemberNames(symbol);
+
+        if (memberNames.Count < 2)
+        {
+            return;
+        }
+
+        var dependencies = new ConcurrentStack<Dependency>();
```

Comment:
Educational: Since the code doesn't care about order, why not `ConcurrentBag`?

Reply:
Because, unlike `ConcurrentStack`, it's not lock-free ([ref](https://devblogs.microsoft.com/pfxteam/faq-are-all-of-the-new-concurrent-collections-lock-free/)). 
This is where we need to be careful with thread contention. The lambda registering on every symbol identifier may be called by N different threads, and using a lock-free collection ensures no thread contention.
That's a general thing that we have been doing with all rules that use `RegisterSymbolStartAction` and register to syntax nodes in that context, to gather data into a single data structure.